0025-nat-traversal.adoc

```diff
@@ -0,0 +1,174 @@
+Peers behind a NAT will be unable to play an active role in the COMIT network.
+
+- + Does not require Bob to configure his NAT
+
+- + Does not require the user to have access to NAT configuration
+
+- + Alice and Bob can  communicate through a direct connection
+
+- - Alice can not dial Bob and therefore cannot initiate a swap or use Bob as a peer to propagate information to the network
+
+- - We will most likely need to redesign protocols to utilise reverse connections
+
+- - Requires Bob to have a known external IP
+
+=== STUN/Hole Punching
+
+Bob sends UDP messages to a third peer known as a STUN server.
+The STUN server will return which ports and public IP Bobs's UDP messages originated from.
+Bob has now discovered public IP and ports which peers can use to dial him.
+Bob can provide this knowledge to Alice, allowing Alice to dial him.
+
+STUN cannot be used to traverse NAT configurations that randomly assign a port and/or IP for new connections.
+For example, a symmetric NAT will randomly generate a new port and/or public IP for each outbound connection made by Alice.
+This prevents Bob from determining or predicting the external IP and ports assigned to her by the NAT.
+
+- + Does not require Bob to configure his NAT
+
+- + Alice and Bob can communicate through a direct connection
+
+- - Cannot traverse symmetric, cone and restricted cone type NATs
+
+- - Requires a third peer to start a connection between Alice and Bob
+
+- - Requires the third peer to know its public IP and ports (either through another STUN/TURN or port forwarding)
+
+- - Requires peers to know peers that are STUN servers
+
+- - Leads to some degree of centralisation
+
+=== TURN
+
+When Bob isn’t able to determine or predict his public facing IP address and port, Bob can dial out to a relay peer, which will keep a long-lived connection open.
+Other peers will be able to dial through the relay peer, which will forward traffic to Bob.
+
+- + Can be used to traverse all commonly know NAT configurations
+
+- + Does not require the Bob to configure his NAT
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires peers to know peers that are TURN servers
+
+- - Leads to some degree of centralisation
+
+== Recommendation
+
+Assumption: A significant portion of comit nodes will have forwarded ports.
+
+We should implement the following NAT traversal methods in the order listed below.
+After each implementation we should evaluate whether it is necessary to implement the subsequent NAT traversal methods.
+
+1. TURN (essential for maximum NAT coverage)
+2. STUN (optional to reduce load and reliance on TURN servers)
```

Comment:
As suggested above, I don't think STUN is a standalone solution.

Reply:
STUN is a standalone solution for NAT traversal. It does not provide complete coverage of every NAT config.